File: app/src/main/java/pl/mareklangiewicz/myhub/io/GitHub.java
package pl.mareklangiewicz.myhub.io;

import java.util.List;

import retrofit2.Call;
import retrofit2.Retrofit;
import retrofit2.converter.moshi.MoshiConverterFactory;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.Headers;
import retrofit2.http.Path;
import rx.Observable;

public final class GitHub {

//    private static final boolean VERY_VERBOSE = false;

    private GitHub() {
        throw new AssertionError("GitHub class is noninstantiable.");
    }


    public static final String URL = "https://api.github.com";

    @SuppressWarnings("unused")
    public static class Plan {
        public String name; // e.g.: "Medium"
        public Long space;
        public Long private_repos;
        public Long collaborators;
    }

    @SuppressWarnings("unused")
    public static class User {
        public String login;
        public long id;
        public String avatar_url;
        public String gravatar_id; // e.g.: ""
        public String url; // e.g.: "https://api.github.com/users/octocat"
        public String html_url; // e.g.: "https://github.com/octocat"
        public String followers_url; // e.g.: "https://api.github.com/users/octocat/followers"
        public String following_url; // e.g.: "https://api.github.com/users/octocat/following{/other_user}"
        public String gists_url; // e.g.: "https://api.github.com/users/octocat/gists{/gist_id}"
        public String starred_url; // e.g.: "https://api.github.com/users/octocat/starred{/owner}{/repo}"
        public String subscriptions_url; // e.g.: "https://api.github.com/users/octocat/subscriptions"
        public String organizations_url; // e.g.: "https://api.github.com/users/octocat/orgs"
        public String repos_url; // e.g.: "https://api.github.com/users/octocat/repos"
        public String events_url; // e.g.: "https://api.github.com/users/octocat/events{/privacy}"
        public String received_events_url; // e.g.: "https://api.github.com/users/octocat/received_events"
        public String type; // e.g.: "User"
        public Boolean site_admin;
        public String name; // e.g.: "monalisa octocat"
        public String company; // e.g.: "GitHub"
        public String blog; // e.g.: "https://github.com/blog"
        public String location; // e.g.: "San Francisco"
        public String email; // e.g.: "octocat@github.com"
        public Boolean hireable;
        public String bio; // e.g.: "There once was..."
        public Long public_repos;
        public Long public_gists;
        public Long followers;
        public Long following;
        public String created_at; // e.g.: "2008-01-14T04:33:35Z"
        public String updated_at; // e.g.: "2008-01-14T04:33:35Z"
        public Long total_private_repos;
        public Long owned_private_repos;
        public Long private_gists;
        public Long disk_usage;
        public Long collaborators;
        public Plan plan;
    }

    @SuppressWarnings("unused")
    public static class Owner {
        public String login; // e.g.: "octocat"
        public Long id;
        public String avatar_url; // e.g.: "https://github.com/images/error/octocat_happy.gif"
        public String gravatar_id; // e.g.: ""
        public String url; // e.g.: "https://api.github.com/users/octocat"
        public String html_url; // e.g.: "https://github.com/octocat"
        public String followers_url; // e.g.: "https://api.github.com/users/octocat/followers"
        public String following_url; // e.g.: "https://api.github.com/users/octocat/following{/other_user}"
        public String gists_url; // e.g.: "https://api.github.com/users/octocat/gists{/gist_id}"
        public String starred_url; // e.g.: "https://api.github.com/users/octocat/starred{/owner}{/repo}"
        public String subscriptions_url; // e.g.: "https://api.github.com/users/octocat/subscriptions"
        public String organizations_url; // e.g.: "https://api.github.com/users/octocat/orgs"
        public String repos_url; // e.g.: "https://api.github.com/users/octocat/repos"
        public String events_url; // e.g.: "https://api.github.com/users/octocat/events{/privacy}"
        public String received_events_url; // e.g.: "https://api.github.com/users/octocat/received_events"
        public String type; // e.g.: "User"
        public Boolean site_admin;

    }

    @SuppressWarnings("unused")
    public static class Permissions {
        public Boolean admin;
        public Boolean push;
        public Boolean pull;
    }

    @SuppressWarnings("unused")
    public static class Repository {
        public Long id;
        Owner owner;
        public String name; // e.g.: "Hello-World"
        public String full_name; // e.g.: "octocat/Hello-World"
        public String description; // e.g.: "This your first repo!"
        // public Boolean private; FIXME SOMEDAY: private is a java keyword too.. check in moshi what can we do..
        public Boolean fork;
        public String url; // e.g.: "https://api.github.com/repos/octocat/Hello-World"
        public String html_url; // e.g.: "https://github.com/octocat/Hello-World"
        public String archive_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/{archive_format}{/ref}"
        public String assignees_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/assignees{/user}"
        public String blobs_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/git/blobs{/sha}"
        public String branches_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/branches{/branch}"
        public String clone_url; // e.g.: "https://github.com/octocat/Hello-World.git"
        public String collaborators_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/collaborators{/collaborator}"
        public String comments_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/comments{/number}"
        public String commits_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/commits{/sha}"
        public String compare_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/compare/{base}...{head}"
        public String contents_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/contents/{+path}"
        public String contributors_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/contributors"
        public String downloads_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/downloads"
        public String events_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/events"
        public String forks_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/forks"
        public String git_commits_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/git/commits{/sha}"
        public String git_refs_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/git/refs{/sha}"
        public String git_tags_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/git/tags{/sha}"
        public String git_url; // e.g.: "git:github.com/octocat/Hello-World.git"
        public String hooks_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/hooks"
        public String issue_comment_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/issues/comments{/number}"
        public String issue_events_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/issues/events{/number}"
        public String issues_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/issues{/number}"
        public String keys_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/keys{/key_id}"
        public String labels_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/labels{/name}"
        public String languages_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/languages"
        public String merges_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/merges"
        public String milestones_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/milestones{/number}"
        public String mirror_url; // e.g.: "git:git.example.com/octocat/Hello-World"
        public String notifications_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/notifications{?since, all, participating}"
        public String pulls_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/pulls{/number}"
        public String releases_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/releases{/id}"
        public String ssh_url; // e.g.: "git@github.com:octocat/Hello-World.git"
        public String stargazers_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/stargazers"
        public String statuses_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/statuses/{sha}"
        public String subscribers_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/subscribers"
        public String subscription_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/subscription"
        public String svn_url; // e.g.: "https://svn.github.com/octocat/Hello-World"
        public String tags_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/tags"
        public String teams_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/teams"
        public String trees_url; // e.g.: "http://api.github.com/repos/octocat/Hello-World/git/trees{/sha}"
        public String homepage; // e.g.: "https://github.com"
        public String language; // e.g.: null
        public Long forks_count;
        public Long stargazers_count;
        public Long watchers_count;
        public Long size;
        public String default_branch; // e.g.: "master"
        public Long open_issues_count;
        public Boolean has_issues;
        public Boolean has_wiki;
        public Boolean has_pages;
        public Boolean has_downloads;
        public String pushed_at; // e.g.: "2011-01-26T19:06:43Z"
        public String created_at; // e.g.: "2011-01-26T19:01:12Z"
        public String updated_at; // e.g.: "2011-01-26T19:14:43Z"
        Permissions permissions;
    }

    private static final String ACCEPT = "Accept: application/vnd.github.v3+json";
    private static final String AGENT = "User-Agent: MyHub";

    public interface Service {

        @Headers({ACCEPT, AGENT})
        @GET("/users/{user}")
        Call<User> getUserCall(@Path("user") String user);

        @Headers({ACCEPT, AGENT})
        @GET("/user")
        Call<User> getUserAuthCall(@Header("Authorization") String auth);

        @Headers({ACCEPT, AGENT})
        @GET("/user")
        Call<User> getUserTFACall(@Header("Authorization") String auth, @Header("X-GitHub-OTP") String code);

        @Headers({ACCEPT, AGENT})
        @GET("/users/{user}/repos?per_page=100")
        Call<List<Repository>> getUserReposCall(@Path("user") String user);

        @Headers({ACCEPT, AGENT})
        @GET("/user/repos?per_page=100")
        Call<List<Repository>> getUserReposAuthCall(@Header("Authorization") String auth);

        @Headers({ACCEPT, AGENT})
        @GET("/user/repos?per_page=100")
        Call<List<Repository>> getUserReposTFACall(@Header("Authorization") String auth, @Header("X-GitHub-OTP") String code);


        @Headers({ACCEPT, AGENT})
        @GET("/users/{user}")
        Observable<User> getUserObservable(@Path("user") String user);

        @Headers({ACCEPT, AGENT})
        @GET("/user")
        Observable<User> getUserAuthObservable(@Header("Authorization") String auth);

        @Headers({ACCEPT, AGENT})
        @GET("/user")
        Observable<User> getUserTFAObservable(@Header("Authorization") String auth, @Header("X-GitHub-OTP") String code);

        @Headers({ACCEPT, AGENT})
        @GET("/users/{user}/repos?per_page=100")
        Observable<List<Repository>> getUserReposObservable(@Path("user") String user);

        @Headers({ACCEPT, AGENT})
        @GET("/user/repos?per_page=100")
        Observable<List<Repository>> getUserReposAuthObservable(@Header("Authorization") String auth);

        @Headers({ACCEPT, AGENT})
        @GET("/user/repos?per_page=100")
        Observable<List<Repository>> getUserReposTFAObservable(@Header("Authorization") String auth, @Header("X-GitHub-OTP") String code);
    }

    private static final Retrofit retrofit = new Retrofit.Builder()
            .baseUrl(URL)
            .addConverterFactory(MoshiConverterFactory.create())
            .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
            .build();

    public static Service create() {


//        if(VERY_VERBOSE) {
//            OkHttpClient client = new OkHttpClient();
//            HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
//            interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
//            client.interceptors().add(interceptor);
//
//            Retrofit loggingretrofit = new Retrofit.Builder()
//                    .baseUrl(URL)
//                    .client(client)
//                    .addConverterFactory(MoshiConverterFactory.create())
//                    .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
//                    .build();
//
//            return loggingretrofit.create(Service.class);
//        }


        return retrofit.create(Service.class);
    }

}

File: app/src/main/java/pl/mareklangiewicz/myhub/io/GitHub.kt
package pl.mareklangiewicz.myhub.io

import retrofit2.Call
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory
import retrofit2.http.GET
import retrofit2.http.Header
import retrofit2.http.Headers
import retrofit2.http.Path
import rx.Observable

//    private static final boolean VERY_VERBOSE = false;

const val URL = "https://api.github.com"

const val ACCEPT = "Accept: application/vnd.github.v3+json"
const val AGENT = "User-Agent: MyHub"



class Plan {
    var name: String? = null // e.g.: "Medium"
    var space: Long? = null
    var private_repos: Long? = null
    var collaborators: Long? = null
}

class User {
    var login: String? = null
    var id: Long = 0
    var avatar_url: String? = null
    var gravatar_id: String? = null // e.g.: ""
    var url: String? = null // e.g.: "https://api.github.com/users/octocat"
    var html_url: String? = null // e.g.: "https://github.com/octocat"
    var followers_url: String? = null // e.g.: "https://api.github.com/users/octocat/followers"
    var following_url: String? = null // e.g.: "https://api.github.com/users/octocat/following{/other_user}"
    var gists_url: String? = null // e.g.: "https://api.github.com/users/octocat/gists{/gist_id}"
    var starred_url: String? = null // e.g.: "https://api.github.com/users/octocat/starred{/owner}{/repo}"
    var subscriptions_url: String? = null // e.g.: "https://api.github.com/users/octocat/subscriptions"
    var organizations_url: String? = null // e.g.: "https://api.github.com/users/octocat/orgs"
    var repos_url: String? = null // e.g.: "https://api.github.com/users/octocat/repos"
    var events_url: String? = null // e.g.: "https://api.github.com/users/octocat/events{/privacy}"
    var received_events_url: String? = null // e.g.: "https://api.github.com/users/octocat/received_events"
    var type: String? = null // e.g.: "User"
    var site_admin: Boolean? = null
    var name: String? = null // e.g.: "monalisa octocat"
    var company: String? = null // e.g.: "GitHub"
    var blog: String? = null // e.g.: "https://github.com/blog"
    var location: String? = null // e.g.: "San Francisco"
    var email: String? = null // e.g.: "octocat@github.com"
    var hireable: Boolean? = null
    var bio: String? = null // e.g.: "There once was..."
    var public_repos: Long? = null
    var public_gists: Long? = null
    var followers: Long? = null
    var following: Long? = null
    var created_at: String? = null // e.g.: "2008-01-14T04:33:35Z"
    var updated_at: String? = null // e.g.: "2008-01-14T04:33:35Z"
    var total_private_repos: Long? = null
    var owned_private_repos: Long? = null
    var private_gists: Long? = null
    var disk_usage: Long? = null
    var collaborators: Long? = null
    var plan: Plan? = null
}

class Owner {
    var login: String? = null // e.g.: "octocat"
    var id: Long? = null
    var avatar_url: String? = null // e.g.: "https://github.com/images/error/octocat_happy.gif"
    var gravatar_id: String? = null // e.g.: ""
    var url: String? = null // e.g.: "https://api.github.com/users/octocat"
    var html_url: String? = null // e.g.: "https://github.com/octocat"
    var followers_url: String? = null // e.g.: "https://api.github.com/users/octocat/followers"
    var following_url: String? = null // e.g.: "https://api.github.com/users/octocat/following{/other_user}"
    var gists_url: String? = null // e.g.: "https://api.github.com/users/octocat/gists{/gist_id}"
    var starred_url: String? = null // e.g.: "https://api.github.com/users/octocat/starred{/owner}{/repo}"
    var subscriptions_url: String? = null // e.g.: "https://api.github.com/users/octocat/subscriptions"
    var organizations_url: String? = null // e.g.: "https://api.github.com/users/octocat/orgs"
    var repos_url: String? = null // e.g.: "https://api.github.com/users/octocat/repos"
    var events_url: String? = null // e.g.: "https://api.github.com/users/octocat/events{/privacy}"
    var received_events_url: String? = null // e.g.: "https://api.github.com/users/octocat/received_events"
    var type: String? = null // e.g.: "User"
    var site_admin: Boolean? = null

}

class Permissions {
    var admin: Boolean? = null
    var push: Boolean? = null
    var pull: Boolean? = null
}

class Repository {
    var id: Long? = null
    internal var owner: Owner? = null
    var name: String? = null // e.g.: "Hello-World"
    var full_name: String? = null // e.g.: "octocat/Hello-World"
    var description: String? = null // e.g.: "This your first repo!"
    // public Boolean private; FIXME SOMEDAY: private is a java keyword too.. check in moshi what can we do..
    var fork: Boolean? = null
    var url: String? = null // e.g.: "https://api.github.com/repos/octocat/Hello-World"
    var html_url: String? = null // e.g.: "https://github.com/octocat/Hello-World"
    var archive_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/{archive_format}{/ref}"
    var assignees_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/assignees{/user}"
    var blobs_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/git/blobs{/sha}"
    var branches_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/branches{/branch}"
    var clone_url: String? = null // e.g.: "https://github.com/octocat/Hello-World.git"
    var collaborators_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/collaborators{/collaborator}"
    var comments_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/comments{/number}"
    var commits_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/commits{/sha}"
    var compare_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/compare/{base}...{head}"
    var contents_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/contents/{+path}"
    var contributors_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/contributors"
    var downloads_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/downloads"
    var events_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/events"
    var forks_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/forks"
    var git_commits_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/git/commits{/sha}"
    var git_refs_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/git/refs{/sha}"
    var git_tags_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/git/tags{/sha}"
    var git_url: String? = null // e.g.: "git:github.com/octocat/Hello-World.git"
    var hooks_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/hooks"
    var issue_comment_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/issues/comments{/number}"
    var issue_events_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/issues/events{/number}"
    var issues_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/issues{/number}"
    var keys_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/keys{/key_id}"
    var labels_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/labels{/name}"
    var languages_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/languages"
    var merges_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/merges"
    var milestones_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/milestones{/number}"
    var mirror_url: String? = null // e.g.: "git:git.example.com/octocat/Hello-World"
    var notifications_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/notifications{?since, all, participating}"
    var pulls_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/pulls{/number}"
    var releases_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/releases{/id}"
    var ssh_url: String? = null // e.g.: "git@github.com:octocat/Hello-World.git"
    var stargazers_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/stargazers"
    var statuses_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/statuses/{sha}"
    var subscribers_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/subscribers"
    var subscription_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/subscription"
    var svn_url: String? = null // e.g.: "https://svn.github.com/octocat/Hello-World"
    var tags_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/tags"
    var teams_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/teams"
    var trees_url: String? = null // e.g.: "http://api.github.com/repos/octocat/Hello-World/git/trees{/sha}"
    var homepage: String? = null // e.g.: "https://github.com"
    var language: String? = null // e.g.: null
    var forks_count: Long? = null
    var stargazers_count: Long? = null
    var watchers_count: Long? = null
    var size: Long? = null
    var default_branch: String? = null // e.g.: "master"
    var open_issues_count: Long? = null
    var has_issues: Boolean? = null
    var has_wiki: Boolean? = null
    var has_pages: Boolean? = null
    var has_downloads: Boolean? = null
    var pushed_at: String? = null // e.g.: "2011-01-26T19:06:43Z"
    var created_at: String? = null // e.g.: "2011-01-26T19:01:12Z"
    var updated_at: String? = null // e.g.: "2011-01-26T19:14:43Z"
    internal var permissions: Permissions? = null
}

interface GithubService {

    @Headers(ACCEPT, AGENT)
    @GET("/users/{user}")
    fun getUserCall(@Path("user") user: String): Call<User>

    @Headers(ACCEPT, AGENT)
    @GET("/user")
    fun getUserAuthCall(@Header("Authorization") auth: String): Call<User>

    @Headers(ACCEPT, AGENT)
    @GET("/user")
    fun getUserTFACall(@Header("Authorization") auth: String, @Header("X-GitHub-OTP") code: String): Call<User>

    @Headers(ACCEPT, AGENT)
    @GET("/users/{user}/repos?per_page=100")
    fun getUserReposCall(@Path("user") user: String): Call<List<Repository>>

    @Headers(ACCEPT, AGENT)
    @GET("/user/repos?per_page=100")
    fun getUserReposAuthCall(@Header("Authorization") auth: String): Call<List<Repository>>

    @Headers(ACCEPT, AGENT)
    @GET("/user/repos?per_page=100")
    fun getUserReposTFACall(@Header("Authorization") auth: String, @Header("X-GitHub-OTP") code: String): Call<List<Repository>>


    @Headers(ACCEPT, AGENT)
    @GET("/users/{user}")
    fun getUserObservable(@Path("user") user: String): Observable<User>

    @Headers(ACCEPT, AGENT)
    @GET("/user")
    fun getUserAuthObservable(@Header("Authorization") auth: String): Observable<User>

    @Headers(ACCEPT, AGENT)
    @GET("/user")
    fun getUserTFAObservable(@Header("Authorization") auth: String, @Header("X-GitHub-OTP") code: String): Observable<User>

    @Headers(ACCEPT, AGENT)
    @GET("/users/{user}/repos?per_page=100")
    fun getUserReposObservable(@Path("user") user: String): Observable<List<Repository>>

    @Headers(ACCEPT, AGENT)
    @GET("/user/repos?per_page=100")
    fun getUserReposAuthObservable(@Header("Authorization") auth: String): Observable<List<Repository>>

    @Headers(ACCEPT, AGENT)
    @GET("/user/repos?per_page=100")
    fun getUserReposTFAObservable(@Header("Authorization") auth: String, @Header("X-GitHub-OTP") code: String): Observable<List<Repository>>
}


private val retrofit = Retrofit.Builder().baseUrl(URL).addConverterFactory(MoshiConverterFactory.create()).addCallAdapterFactory(RxJavaCallAdapterFactory.create()).build()

fun createGithubService(): GithubService {


        // THIS IS OLD JAVA WAY TO INJECT SOME REQUEST/RESPONSE LOGGING
        //        if(VERY_VERBOSE) {
        //            OkHttpClient client = new OkHttpClient();
        //            HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
        //            interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
        //            client.interceptors().add(interceptor);
        //
        //            Retrofit loggingretrofit = new Retrofit.Builder()
        //                    .baseUrl(URL)
        //                    .client(client)
        //                    .addConverterFactory(MoshiConverterFactory.create())
        //                    .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
        //                    .build();
        //
        //            return loggingretrofit.create(Service.class);
        //        }


    return retrofit.create(GithubService::class.java)
}

